docs/proposals/20210809-containerpriority.md

```diff
@@ -0,0 +1,126 @@
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+
+Containers in a same Pod in it might have dependence, which means the application in one container runs depending on another container.
+For example:
+
+1. Container A has to start first. Container B can start only if A is already running.
+2. Container B has to exit first. Container A can stop only if B has already exited.
+
+Currently, the sequences of containers start and stop are controlled by Kubelet.
+Kubernetes used to have a [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/753-sidecar-containers),
+which plans to add a type field for container to identify the priority of start and stop.
+However, it has been refused because of [sig-node thought it may bring a huge change to code](https://github.com/kubernetes/enhancements/issues/753#issuecomment-713471597).
+
+Now, OpenKruise tries to provide a way to help users control the sequence of containers start in a Pod.
+Unfortunately, the sequence of stop can still only be controlled by the preStop hook in containers, which requires user to configure.
+
+## Proposal
+
+### API definition
+
+Users can set the priority in `env` of each container:
+
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: main
+    # ...
+  - name: sidecar
+    env:
+    - name: KRUISE_CONTAINER_PRIORITY
+      value: "1"
+    # ...
+```
+
+1. The range of the value is `[-2147483647, 2147483647]`. Defaults to `0` if no such env exists.
+2. The container with higher priority will be guaranteed to start before the others with lower priority.
+3. The containers with same priority have no limit to their start sequence.
+4. The priority env can also be used to control the sequence of init containers executing.
+
+*Note that the Pod has no limit to its owner, which means it can be created by Deployment, CloneSet or any other controllers.*
+
+### Implementation Details/Notes/Constraints
+
+When Kruise webhook find `KRUISE_CONTAINER_PRIORITY` env in a creating Pod, it should inject `KRUISE_CONTAINER_BARRIER` env into containers.
+
+The value of `KRUISE_CONTAINER_BARRIER` is set from a ConfigMap named `{pod-name}-barrier` and the key is related to priority of this container.
+
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: main
+    # ...
+    env:
+    - name: KRUISE_CONTAINER_BARRIER
+      valueFrom:
+        configMapKeyRef:
+          name: {pod-name}-barrier
+          key: "c_0"   # no KRUISE_CONTAINER_PRIORITY, so defaults to 0
+  - name: sidecar
+    env:
+    - name: KRUISE_CONTAINER_PRIORITY
+      value: "1"
+    - name: KRUISE_CONTAINER_BARRIER
+      valueFrom:
+        configMapKeyRef:
+          name: {pod-name}-barrier
+          key: "c_1"   # its KRUISE_CONTAINER_PRIORITY is 1
+    # ...
+```
+
+*Note that the key of containers will be `c_{priority}` and the key of init containers will be `i_{priority}`.*
```

Comment:
the start sequence of init container is already deterministic, i don't think it is necessary to mess with the init containers

Reply:
Removed, don't need to deal with init containers.